### DISCUSSION TITLE
Is it possible to access the frontMatter from included partial files?

### DISCUSSION THREAD

**Discussion by ilg-ul** [facebook/docusaurus#8785] [Q&A]
I moved some content from a docs page to a partial file, and I noticed that the references to the frontMatter are empty.

I understand that the included pages are separate components, each with their environment, but it would be nice to have access to the parent page's frontMatter.

If this is not possible, is there any other mechanism to pass data from parent to included partial files? Or I have to explicitly define attributes and pass them as params?

Thank you!

**Comment by pranabdas** [comment]
Yes, you need to pass `frontMatter` object to `<Partial />` component as a prop and then you can access it from partial.

Primary markdown file:
```
---
version: 3.4
---

import Partial from './_part.mdx';

<Partial frontMatter={frontMatter} />
```

Now `_part.mdx` can use `frontMatter`:
```
<p>Version: {props.frontMatter.version}</p>
```

**Reply by ilg-ul** [reply]
Yeah, that must be functional, but looks like an ugly hack :-(

Instead of passing the frontMatter, I'll probably pass the variables themselves.

@slorber, what do you think about the frontMatter being propagated automatically when importing partials?

**Reply by slorber** [reply]
> Yeah, that must be functional, but looks like an ugly hack :-(

No that's how MDX works: MDX files are compiled to React components, and we pass things as props to React components.

> Instead of passing the frontMatter, I'll probably pass the variables themselves.

You are free to use the props API to want for your React components (MDX files included)

> @slorber, what do you think about the frontMatter being propagated automatically when importing partials?

We don't plan to do that.

A partial might be used in multiple places at once. It can be imported in one file that has frontMatter, and another that doesn't (for example your React landing page can import that partial => no frontMatter). 

Really you should take the mental model that an MDX file is a React component. FrontMatter (and any other prop) is not automatically "propagated" in React components, so we don't do it either for MDX files which are React components.


---


Apart props, there's another way to "propagate" data to a whole subtree of React components: the React context. This project will make the frontMatter accessible to a whole subtree of React components (including MDX files but also JS/TSX files): https://github.com/roydukkey/docusaurus-theme-frontmatter => we are likely to add this feature in core someday.

If you are sure your partial is always used in a context where frontmatter is available, you should be able to do:

```mdx
import useFrontMatter from '@theme/useFrontMatter';

# Markdown file

Hello {useFrontMatter().name}
```

**Reply by ilg-ul** [reply]
Ok, from a React point of view this makes sense; so a bit more extra code to make it work. 

> Really you should take the mental model that an MDX file is a React component.

I'm doing my best, but the React learning curve is steep, and it'll take me some time, in Jekyll things were completely different (and much easier to implement). Plus that Web development in general is not part of my usual work.

The [docusaurus-theme-frontmatter](https://github.com/roydukkey/docusaurus-theme-frontmatter) project is interesting, I'll probably try to evaluate it, but my first choice would probably be to pass the variables directly, it seems a more generic solution (and probably closer to the React component spirit, isn't it @slorber?).

**Reply by slorber** [reply]
MDX are React components with props.

You can process Markdown with Remark plugin for what you call "variables". We don't have opinions on which remark plugins you decide to use but it's your responsibility.

I don't know what to recommend you, you need to understand the tradeoffs of different possible approaches and make a choice between "placeholders" that are substituted at build time with Remark, or real runtime variables used in React components. Both are different things, and we enable you to use both.

**Reply by ilg-ul** [reply]
> MDX are React components with props. 

Yes, probably props is the correct term, not variables.

> You can process Markdown with Remark plugin

I don't know yet how to do this, thus it seems complicated.

I was thinking on a simple solution, like:

```
---
title: How to install the xPack Docus Mock binaries
version: "1.11.0"
---

import InstallWindows from './_install-windows.mdx';

## Overview
...

<InstallWindows version={frontMatter.version} />
```

and in `_install-windows.mdx`

```
...
< pre >
C:\>%USERPROFILE%\AppData\Roaming\xPacks\@xpack-dev-tools\docus-mock\{ props.version }-1.1\.content\bin\docus-mock.exe --version
<br/>
{ props.version }
</pre >
```

@slorber, is this closer to the spirit of React/MDX/Docusaurus?

**Reply by slorber** [reply]
@ilg-ul your suggestion is how we use props in REact, and an MDX file is a React component. So yes, that works. There's no single solution here.

> I don't know yet how to do this, thus it seems complicated.

If you want to use and create non-React/Markdown std features that's worth learning Remark, but yes there's a learning curve
